File: packages/shared/src/types/stats.ts
/**
 * Types for the My Stats screen payload (GET /api/users/me/stats).
 *
 * Mirror the Python Pydantic schemas in apps/backend/models/schemas.py.
 * Shape-contract tests in the backend test suite enforce that these two
 * stay in sync.
 */

export interface EloTimelinePoint {
  readonly date: string;
  readonly rating: number;
}

export interface MyStatsOverall {
  readonly wins: number;
  readonly losses: number;
  readonly games_played: number;
  readonly rating: number;
  readonly peak_rating: number;
  readonly win_rate: number;
  readonly current_streak: number;
  readonly avg_point_diff: number;
}

export interface MyStatsTrophy {
  readonly league_id: number;
  readonly league_name: string;
  readonly season_name: string;
  readonly place: number;
}

/** Stats row for a single partner or opponent relationship. No rating_diff. */
export interface MyStatsRelationStat {
  readonly player_id: number;
  readonly display_name: string;
  readonly initials: string;
  readonly games_played: number;
  readonly wins: number;
  readonly losses: number;
  readonly win_rate: number;
}

export interface MyStatsPayload {
  readonly player_name: string;
  readonly player_city: string | null;
  readonly player_level: string | null;
  readonly overall: MyStatsOverall;
  readonly trophies: readonly MyStatsTrophy[];
  readonly partners: readonly MyStatsRelationStat[];
  readonly opponents: readonly MyStatsRelationStat[];
  readonly elo_timeline: readonly EloTimelinePoint[];
}

File: apps/backend/models/schemas/stats.py
"""Stats models."""

from typing import Optional, List
from pydantic import BaseModel, ConfigDict


class EloTimelinePoint(BaseModel):
    """Single ELO rating data point for a player's rating history chart."""

    date: str
    rating: int


class MyStatsOverall(BaseModel):
    """Overall aggregate stats for the authenticated player."""

    wins: int
    losses: int
    games_played: int
    rating: int
    peak_rating: int
    win_rate: float
    current_streak: int
    avg_point_diff: float


class MyStatsTrophy(BaseModel):
    """A single season placement trophy (gold/silver/bronze only)."""

    league_id: int
    league_name: str
    season_name: str
    place: int


class MyStatsRelationStat(BaseModel):
    """Stats for a single partner or opponent relationship."""

    player_id: int
    display_name: str
    initials: str
    avatar_url: Optional[str] = None
    games_played: int
    wins: int
    losses: int
    win_rate: float


class MyStatsPayload(BaseModel):
    """Full my-stats payload returned by GET /api/users/me/stats."""

    model_config = ConfigDict(extra="ignore")

    player_name: str
    player_avatar_url: Optional[str] = None
    player_city: Optional[str] = None
    player_level: Optional[str] = None
    overall: MyStatsOverall
    trophies: List[MyStatsTrophy]
    partners: List[MyStatsRelationStat]
    opponents: List[MyStatsRelationStat]
    elo_timeline: List[EloTimelinePoint]


class SeasonAwardResponse(BaseModel):
    """Season award data for API responses."""

    model_config = ConfigDict(extra="ignore")

    id: int
    season_id: int
    season_name: Optional[str] = None
    league_id: int
    league_name: Optional[str] = None
    player_id: int
    player_name: Optional[str] = None
    player_avatar: Optional[str] = None
    player_profile_picture_url: Optional[str] = None
    award_type: str
    award_key: str
    rank: Optional[int] = None
    value: Optional[float] = None
    created_at: Optional[str] = None
